File: java/DateOfBirth.java
// Example 6: DateOfBirth.java

record Day(int day) {
    Day {
        Check.validity(0 < day && day <= 31, day + ": day of month out of range");
    }
}

record Month(int month) {
    Month {
        Check.validity(0 < month && month <= 12, month + ": month out of range");
    }
}

record Year(int year) {
    Year {
        Check.validity(1900 < year && year <= 2022, year + ": year out of range");
    }
}

record BirthDate(Month m, Day d, Year y) {}

public class DateOfBirth {
    public static void main(String[] args) {
        System.out.println(new BirthDate(new Month(7), new Day(8), new Year(1957)));
        System.out.println(new BirthDate(new Month(0), new Day(32), new Year(1857)));
    }
}

File: kotlin/DateOfBirth.kt
// Example 6: DateOfBirth.java
data class Day(val d: Int) {
    init {
        check(0 < && <= 31, . toString () + ": day of month out of range")
    }
}

data class Month(val m: Int) {
    init {
        check(0 < && <= 12, . toString () + ": month out of range")
    }
}

data class Year(val y: Int) {
    init {
        check(1900 < && <= 2022, . toString () + ": year out of range")
    }
}

data class BirthDate(val month: Month, val day: Day, val year: Year)

fun main() {
    System.out.println(BirthDate(Month(7), Day(8), Year(1957)))
    System.out.println(BirthDate(Month(0), Day(32), Year(1857)))
}
